
# GEOG59950 Programming for Geographical Information Analysis: Core Skills <a class="tocSkip">

 #### Contact: F.L.Pontin@leeds.ac.uk <a class="tocSkip">

# Exercise 2: Getting started

- Data types

## Using Jupyter notebooks

Welcome to coding in Python! This course is designed to introduce you to the basics of coding in Python and then get you to apply your new-found coding skills to carry out some spatial data analytics!
This session is written in a Jupyter Notebook. A Jupyter Notebook ["is an open source web application that you can use to create and share documents that contain live code, equations, visualizations, and text"](https://realpython.com/jupyter-notebook-introduction/).

A Jupyter Notebook is made up of cells. These cells can contain text (known as markdown), code, images and many other formats.
 E.g.
To select a cell click on the text or grey cell containing code and a thin grey line will appear around the cell and a small toolbar will appear on the top right-hand side of the cell. <br>

![Select text cell](https://github.com/FrancescaPontin/GEOG5990M/blob/main/screenshots/Select_text_cell_colab.png?raw=1)
![Select code cell](https://github.com/FrancescaPontin/GEOG5990M/blob/main/screenshots/Select_code_cell_colab.png?raw=1)


Have a look at all the functions on the right-hand toolbar ('Move up', 'Move down', 'Copy link to cell', 'Edit', etc...)


To edit a text cell click in it twice, a toolbar will appear at the top of the cell, the left hand-side will show the raw text and right-hand side shows the 'rendered' code.

![Edit text cell](https://github.com/FrancescaPontin/GEOG5990M/blob/main/screenshots/Edit_text_cell_colab.png?raw=1)

During the teaching session you will read through the Jupyter Notebok, running sections of code yourself to learn key data analysis skills. There will also be places to edit and write your own code within the notebook.

* ##### <font color='orchid'>Instructions and tasks for you to complete are in purple</font>
    
* ##### Where you have to write your own code answers are provided at the end of the section

The first few exercises are a recap from last week

## Hello World

Following coding tradition the first thing we are going to program is to print the words "Hello World".

The <code>print() </code> function prints the specified message to your screen

The speech marks <code>" "</code> arround Hello World let python know you are typing text (also known as a string).

 <font color='orchid'> <b>Run the code below</b></font>

In [ ]:
print("Hello World")

## Basic maths

Lets do some basic maths
<br><font color='orchid'> <b>Run the code below</b></font>

In [ ]:
2+2

If we print 2+2 we get the same thing
<br> <font color='orchid'> <b>Run the code below</b></font>

In [ ]:
print(2+2)

If we <code>print("2+2")</code> however we get the characters 2 + 2 as the speach marks tell Python we are typing text and not numerical characters


<br> <font color='orchid'> <b>Run the code below</b></font> and see for yourself

In [ ]:
print("2+2")


## Assigning values to variables

We can also assign the sum 2+2 to a variable.

In this case we have named the variable <font color='blue'>answer</font>. We can now type <font color='blue'>answer</font> instead of 2+2 every time.

<br> <font color='orchid'> <b>To assign 2+2 to the variable answer run the code below</b></font>

In [2]:
answer = 2+2
print(answer)

4


We can also add our created variable to a new equation
<br> <font color='orchid'> <b>To add 5 to the variable answer run the code below</b></font>

In [3]:
# we can also add our created variable to a new equation
answer + 5

9

NOTE: <code>#</code> When a hashtag is added to a line of code the rest of that line is not treated as code.
This can be used to add comments to your code so you know what it is doing (especially useful when you come back to it later on!)

## Lists

Remember lists are an ordered collection of one or more data item.  Defined by square brackets <code>[]</code>

<br> <font color='orchid'> <b>Look back at the lecture notes and create a list called fruit with the following data items: 'apple', 'pear', 'banana', 'strawberry' (note these are strings)</b></font>

In [4]:
# Type code here
fruit=['apple','pear','banana','strawberry']

We can check that you have created a list using the <code>type</code> function

<br> <font color='orchid'> <b>Enter <code> type(fruit) </code> into the cell below</b></font>

In [5]:
# run this code
type(fruit)


list

### Lists and basic maths

We can also create a list of numbers.  <br> <font color='orchid'> <b>Run the code below </b></font>. What happens when we try to multiply the list by 3?

In [7]:
numbers = [1,3,5,7,9,11]

numbers*3

[1, 3, 5, 7, 9, 11, 1, 3, 5, 7, 9, 11, 1, 3, 5, 7, 9, 11]

### List comprehension

To multiply each item in the list (and create a new list of the results) we need to select each value <code>i</code> in the list:
<code>[i*3 for i in numbers]</code>

I.e. for each element 1 to i in the list named numbers, multiply that element by 3

<br> <font color='orchid'> <b> Run the code to see</b></font>

In [8]:
# multiply element by 3, repeat for every element in the list 'numbers'
new_numbers = [i*3 for i in numbers]

# print the new list
print(new_numbers)

[3, 9, 15, 21, 27, 33]


<div class="alert alert-block alert-warning">

### A very quick introduction to for loops

Making a list based on an old list is known as list comprehension. An alternative is a for loop as shown below.

We will come back to these later, so for now just <font color='orchid'> <b> run the code and read the comments explaining what each step does. </b></font>

In [9]:
# create a new empty list
my_new_list = []

# for element in the list 'numbers':
for i in numbers:
    # multiply element by 3 and append the result to the new empty list
    my_new_list.append(i * 3)

# print the new list
print(my_new_list)

[3, 9, 15, 21, 27, 33]


In [16]:
numbers_list=[i*3 for i in numbers]
print(numbers_list)

[3, 9, 15, 21, 27, 33]


We can also multiply by another defined variable. For example we previously defined the variable <code>answer</code> (<code>answer = 2+2</code>)

In [11]:
# multiply element by variable 'answer', repeat for every element in the list 'numbers'
[i*answer for i in numbers]

[4, 12, 20, 28, 36, 44]

Commonly you will see <code>i</code> and <code>j</code> to define elements in a list. However this is just convention and you could use anything you wanted to refer to the elements in the list e.g. <code>elephants</code>

(Though typing <code>i</code> is a lot quicker)

In [10]:
# multiply element by 3, repeat for every element in the list 'numbers'
[elephants*3 for elephants in numbers]

[3, 9, 15, 21, 27, 33]

## First look at data frames

### .head() and .tail() functions
To understand how to explore data frames and different data types in python we are going to use a set of data about passengers on the titanic. This is an example dataset built into the seaborn python package.

We will go into detail about reading in data and loading packages in the next exercise, <font color='orchid'> <b>for now run the cell of code below. </b></font>

Note the <code>.head()</code> function shows the top 5 lines of the data frame.

In [17]:
# Import the seaborn package
import seaborn as sns

# load the titanic example dataset and save it as a dataframe named titanic
titanic = sns.load_dataset('titanic')

# look at the first 5 rows of the dataframe
titanic.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

Note NaN denotes a cell containing no data - a null cell

<font color='orchid'> <b> Try entering and running <code>titanic.tail()</code> instead of <code>.head()</code> in the cell below. </b></font> What view of the dataframe do you think you are now seeing?

In [18]:
# enter the instructed code here
titanic.tail()

survived  pclass     sex   age  sibsp  parch   fare embarked   class  \
886         0       2    male  27.0      0      0  13.00        S  Second   
887         1       1  female  19.0      0      0  30.00        S   First   
888         0       3  female   NaN      1      2  23.45        S   Third   
889         1       1    male  26.0      0      0  30.00        C   First   
890         0       3    male  32.0      0      0   7.75        Q   Third   

       who  adult_male deck  embark_town alive  alone  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True


    
<I> A Quick description of the titanic data variables:
- <b>survival:</b>    If the passenger survived
- <b>PassengerId:</b> Unique Id of a passenger.
- <b>pclass:</b>    Ticket class
- <b>sex:</b>   Sex     
- <b>Age:</b>   Age in years     
- <b>sibsp:</b>    Number of siblings / spouses aboard the Titanic     
- <b>parch:</b>   Number of parents / children aboard the Titanic     
- <b>ticket:</b>   Ticket number     
- <b>fare:</b>   Cost of the passenger fare     
- <b>cabin:</b>  Cabin number     
- <b>embarked:</b>    Port of Embarkation</I>

### Data frame columns

We might just want to get a list of the columns in the dataframe to give us a quick idea of what data we have present. To do this we can use the <code>.columns</code> function after we name the dataframe.

<font color='orchid'> <b> Enter <code>titanic.columns</code> in the cell below. </b></font>. The columns listed should be the same as the columns in the <code>.head()</code> view of the dataframe.


In [19]:
# enter the code here
titanic.columns

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')

#### Referring to a column in a dataframe

If we want to select a single column of the dataframe we can also do that.

There are several ways to refer to a column in a pandas dataframe.

The easiest way is by putting the name of the column in square barckets and speech marks <code>[" "]</code> after the name of the dataframe.
e.g. <code><font color='blue'>dataframe_name</font>["<font color='blue'>column_name</font>"]</code>
    
<font color='orchid'> <b> Try to select the 'fare' column from the titanic dataframe.

In [21]:
# enter the code here
titanic["fare"]

0       7.2500
1      71.2833
2       7.9250
3      53.1000
4       8.0500
        ...   
886    13.0000
887    30.0000
888    23.4500
889    30.0000
890     7.7500
Name: fare, Length: 891, dtype: float64

*Note: To save space only a snapshot of the column is shown and not all the rows

### Data frame index
When we type <code> dataframe_name["column_name"]</code> we get the values of the column but we also see the index (the row names). In this case the rows are just numbered 0:890. But these could be other values such as passenger names.

Python indexing starts at 0 not 1. So the first row is row 0 and the first column column 0.

<code>.index</code> works the same as <code>.columns</code> but this time shows the row names.

<font color='orchid'> <b> Run <code>titanic.index</code> in the cell bellow </font>

In [22]:
# enter the code here
titanic.index

RangeIndex(start=0, stop=891, step=1)

We can see the index starts at 0, stops after 891 enteries and increases by a step of 1 for each row.

### Data frame shape
To get the number of rows and columns of a dataframe we can use <code>.shape</code>

<font color='orchid'> <b> Run <code>titanic.shape</code> in the cell bellow </font>

In [23]:
# enter the code here
titanic.shape

(891, 15)

## Data Types

### Data types recap

As we covered in the lecture there are different types of data:

<b>Objects:</b> also known as strings or written characters/text in plain english e.g. Hello World

<b>Intergers:</b> Whole numbers e.g. 2, 57 or 109567835

<b>Floats:</b> A number with a decimal place e.g. 2.34534, 5.5 or 1.0

<b>Boolean:</b> True or False data type
<br>

<b>Datetime:</b> Values that are either a date, time or both e.g. 2019-10-31 09:26:03.478039 (9:26 am on Halloween 2019)

<b>Category:</b> A fintie list of text values E.g. London, Paris, Berlin, Rome (There are a finite number of captial cities)

Learn more about python data types using this realpython [online resource](https://realpython.com/python-data-types/)
<br>
<br>
<br>

### Checking the data type

We can check the data type of each column in a dataframe using the <code>.dtypes</code> funciton
<br> <font color='orchid'> <b>Run the code below</b></font> and have a look at the data type of each of the columns. Are they all as you expected?

In [24]:
titanic.dtypes

survived          int64
pclass            int64
sex              object
age             float64
sibsp             int64
parch             int64
fare            float64
embarked         object
class          category
who              object
adult_male         bool
deck           category
embark_town      object
alive            object
alone              bool
dtype: object

<code>.info()</code> gives us slightly more information including:
- data types: (<code>.dtypes</code>)
- null counts: number of rows containing non-null values
- memory usage: how much computer memory the table uses (useful to know to stop your code running)

If a column has fewer non-null values than the total number of rows this indicates that data might be missing.
<br> <font color='orchid'> <b>Run the code below</b></font>

In [25]:
titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.7+ KB


## Extra tasks

If you have time at the end of the session run the code below to load the penguins dataset and answer the following question using your new found coding skills.
    
- How many columns and rows does the penguins data frame have?
- What are the data types of the different columns in the penguindataframe?
- Is there any data obviously missing from the dataframe?
- Can you use markdown to produce a list fo the column names and variable types for the penguins data set as I have done above for the titanic data set

Work through the following:
- https://www.programiz.com/python-programming/comments
- https://www.programiz.com/python-programming/variables-constants-literals
- https://www.programiz.com/python-programming/variables-datatypes
- https://www.programiz.com/python-programming/operators

In [26]:
# load the penguins example dataset and save it as a dataframe named penguins
penguins = sns.load_dataset('penguins')

In [28]:
# How many columns and rows does the coffee data frame have? 7,345
penguins.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   species            344 non-null    object 
 1   island             344 non-null    object 
 2   bill_length_mm     342 non-null    float64
 3   bill_depth_mm      342 non-null    float64
 4   flipper_length_mm  342 non-null    float64
 5   body_mass_g        342 non-null    float64
 6   sex                333 non-null    object 
dtypes: float64(4), object(3)
memory usage: 18.9+ KB


In [29]:
# What are the data types of the different columns in the penguins dataframe?
penguins.head()

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0  Adelie  Torgersen            39.1           18.7              181.0   
1  Adelie  Torgersen            39.5           17.4              186.0   
2  Adelie  Torgersen            40.3           18.0              195.0   
3  Adelie  Torgersen             NaN            NaN                NaN   
4  Adelie  Torgersen            36.7           19.3              193.0   

   body_mass_g     sex  
0       3750.0    Male  
1       3800.0  Female  
2       3250.0  Female  
3          NaN     NaN  
4       3450.0  Female

In [34]:
# Is there any data obviously missing from the dataframe?
#nan_rows_sex = penguins[penguins['sex'].isna()]
#print(nan_rows_sex)

nan_rows_mass = penguins[penguins['body_mass_g'].isna()]
print(nan_rows_mass)

    species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
3    Adelie  Torgersen             NaN            NaN                NaN   
339  Gentoo     Biscoe             NaN            NaN                NaN   

     body_mass_g  sex  
3            NaN  NaN  
339          NaN  NaN  


Use the following code to get a list of other dataframes you can explore
<code>sns.get_dataset_names()</code>, can you produce a summary of what the data is showing and idenitfy where data is missing?

You might need to google the dataset to understand the variable names. You might want to create a markdown data summary similar to the one I created for the titanic dataset:

    
<I> A Quick description of the titanic data variables:
- <b>survival:</b>    If the passenger survived
- <b>PassengerId:</b> Unique Id of a passenger.
- <b>pclass:</b>    Ticket class
- <b>sex:</b>   Sex     
- <b>Age:</b>   Age in years     
- <b>sibsp:</b>    Number of siblings / spouses aboard the Titanic     
- <b>parch:</b>   Number of parents / children aboard the Titanic     
- <b>ticket:</b>   Ticket number     
- <b>fare:</b>   Cost of the passenger fare     
- <b>cabin:</b>  Cabin number     
- <b>embarked:</b>    Port of Embarkation</I>

In [38]:
sns.get_dataset_names()

['anagrams',
 'anscombe',
 'attention',
 'brain_networks',
 'car_crashes',
 'diamonds',
 'dots',
 'dowjones',
 'exercise',
 'flights',
 'fmri',
 'geyser',
 'glue',
 'healthexp',
 'iris',
 'mpg',
 'penguins',
 'planets',
 'seaice',
 'taxis',
 'tips',
 'titanic']

In [42]:
diamonds= sns.load_dataset('diamonds')
diamonds.head()


carat      cut color clarity  depth  table  price     x     y     z
0   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
1   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31
2   0.23     Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31
3   0.29  Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63
4   0.31     Good     J     SI2   63.3   58.0    335  4.34  4.35  2.75